service/internal/endpoint-discovery/middleware.go

```diff
@@ -0,0 +1,101 @@
+package endpoint_discovery
+
+import (
+	"context"
+	"fmt"
+	"github.com/aws/aws-sdk-go-v2/aws"
+	awsmiddleware "github.com/aws/aws-sdk-go-v2/aws/middleware"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/smithy-go/logging"
+	smithyhttp "github.com/aws/smithy-go/transport/http"
+
+	"github.com/aws/smithy-go/middleware"
+)
+
+// invalidConfigErrorCode denotes type of config error
+const invalidConfigErrorCode = "Invalid configuration"
+
+// DiscoverEndpointOptions are optionals used with DiscoverEndpoint operation.
+type DiscoverEndpointOptions struct {
+
+	// EndpointUsedForDiscovery is the endpoint used to make an endpoint
+	// discovery call.
+	EndpointUsedForDiscovery string
+
+	// DisableHTTPS will disable tls for endpoint discovery call and
+	// subsequent discovered endpoint if service did not return an
+	// endpoint scheme.
+	DisableHTTPS bool
+}
+
+// DiscoverEndpoint is a serialize step middleware used to discover endpoint
+// for an API operation.
+type DiscoverEndpoint struct {
+
+	// Options provides optional settings used with
+	// Discover Endpoint operation.
+	Options []func(*DiscoverEndpointOptions)
+
+	// DiscoverOperation represents the endpoint discovery operation that
+	// returns an Endpoint or error.
+	DiscoverOperation func(ctx context.Context, key string, required bool, options ...func(*DiscoverEndpointOptions)) (WeightedAddress, error)
+
+	// EndpointDiscoveryEnableState represents the customer configuration for endpoint
+	// discovery feature.
+	EndpointDiscoveryEnableState config.EndpointDiscoveryEnableState
```

Comment:
`EndpointDiscoveryEnableState` should be moved to `aws` since nothing in the SDK may depend on the `config` module/pkg.